contracts/contracts/staking/SingleAssetStaking.sol

```diff
@@ -0,0 +1,205 @@
+    }
+
+    function _supportedDuration(uint256 duration) internal view returns (bool) {
+      for (uint i =0; i < durations.length; i++) {
+        if (duration == durations[i]) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+
+
+    /* ========== VIEWS ========== */
+
+    function totalStaked(address account) external view returns (uint256 total) {
+      Stake[] storage stakes = userStakes[account];
+
+      for (uint i=0; i< stakes.length; i++) {
+        total += stakes[i].amount;
+      }
+    }
+
+    function totalExpectedRewards(address account) external view returns (uint256) {
+      return _totalExpectedRewards(userStakes[account]);
+    }
+
+    function totalCurrentHoldings(address account) external view returns (uint256 total) {
+      Stake[] storage stakes = userStakes[account];
+
+      for (uint i=0; i< stakes.length; i++) {
+        Stake storage stake = stakes[i];
+        if (stake.end < block.timestamp) {
+          total += _totalExpected(stake);
+        } else {
+          //calcualte the precentage accrued in term of rewards
+          total += stake.amount
+          .add(
+            stake.amount
+            .mulTruncate(
+              _calcRewardMultiplier(stake.rate, stake.duration)
+            ).mulTruncate( 
+            stake.duration.sub(stake.end.sub(block.timestamp))
+            .divPrecisely(stake.duration)
+                         )
+          );
+        }
+      }
+    }
+
+    /* ========== MUTATIVE FUNCTIONS ========== */
+
+    function stake(uint256 amount, uint256 duration) external {
+        require(!paused, "Staking paused");
+        require(amount > 0, "Cannot stake 0");
+        require(_supportedDuration(duration), "Invalid duration");
+
+        Stake[] storage stakes = userStakes[msg.sender];
+        
+        uint256 end = block.timestamp + duration;
+
+        uint i = stakes.length; // start counting at the end of the current array
+        stakes.length += 1;     //grow the array;
+        // find the spot where we can insert the current stake
+        while (i != 0  && stakes[i-1].end < end) {
+          // shift it back one
+          stakes[i] = stakes[i-1];
+          i -= 1;
+        }
+
+        // insert the stake
+        Stake storage newStake = stakes[i];
+        newStake.rate = rewardRate;
+        newStake.end = end;
+        newStake.duration = duration;
+        newStake.amount = amount;
+
+        stakingToken.safeTransferFrom(msg.sender, address(this), amount);
+        totalOutstanding = totalOutstanding.add(_totalExpected(newStake));
+        //we need to have enough balance to cover the total outstanding after this.
+        require(stakingToken.balanceOf(address(this)) >= totalOutstanding, "Insufficient rewards");
+        emit Staked(msg.sender, amount);
+    }
+
+    function exit() external  {
+        Stake[] storage stakes = userStakes[msg.sender];
```

Comment:
If paused should exits be paused as well?

Reply:
@micahalcorn ?